Some more getters

ralsei opened the pull request:
This adds a few utility functions, and therefore closes #8.

Changes to existing functions:
- `bsearch` now has a `#:rightmost?` argument, which determines whether to get the index of the rightmost element in the vector, rather than the leftmost one. necessary for `series-index-range`

New functions:
- `series-index-range`: binary searches on a series, and returns both the left and rightmost index of an element. useful for when a series has duplicates
- `df-index-range`: same for data-frames
- `series-is-sorted?`: determines if a series has a comparison function. useful for knowing whether `series-index-of` is possible, or sorting/linear search is required
- `df-is-sorted?`: same for data-frames
- `series-shallow-copy`: copies a series's vector, retaining all of its properties
- `df-duplicate-series`: returns a copy of the given series from the data-frame, retaining all of its properties, using `series-shallow-copy`

Nit-picks:
- added `df-shallow-copy` to docs
- fixed `for/data-frame` syntax in docs (call that cleaning up my own mess)

All functions have tests, but do feel free to let me know if they're insufficient.

ralsei committed:
add df-index-range, for when a series has dups

ralsei committed:
add df-is-sorted?

designed for checking if `df-index-of` is viable, or linear search
or sorting is necessary first

ralsei committed:
add df-duplicate-series

ralsei committed:
fix for/data-frame syntax docs

alex-hhh commented on private/df.rkt:
```diff
@@ -267,6 +280,14 @@
+;; If SERIES is not sorted, this will raise an error. This will return a pair of the
+;; leftmost and rightmost index in the given series, in case the value is duplicated.
+;; All other behavior is similar to or the same as df-index-of.
+(define (df-index-range df series value)
```
I think `df-equal-range` would be a better name for this function.  Similarily, `series-equal-range` for the series one...

alex-hhh commented on private/series.rkt:
```diff
@@ -156,6 +156,13 @@
+;; Creates a copy of the series C. The returned copy will reference the same
+;; elements as the original, but any add/delete operations and blessings
+;; will only affect the copy.
+(define (series-shallow-copy c)
```
This is actually a "deep" copy of the series, since you are copying the contents of the `data` vector.  A shallow copy would just copy the reference to the vector.  You can rename this to `series-copy`, or (my preference) `copy-series`.

Also, you could implement this in terms of `struct-copy` since everything is just a plain copy except for the data vector.   This has the advantage that the code will not have to change in the unlikely event that we add additional slots to the series struct


```racket
(struct-copy series c [data (vector-copy (series-data c))])
```

alex-hhh commented on private/bsearch.rkt:
```diff
@@ -58,6 +62,25 @@
                   (do-search (+ mid 1) end)
                   mid)))))
 
+  (define (do-reverse-search start end)
```
Looking at the `bsearch` implementation it seems to me that it is not well suited to handling series which contains equal values.  I wanted it to exit early if the search happens to land on the value searched for, but this optimization does not play well if the series contains equal values. In particular, it requires using weak ordering, i.e. `<=` to find the actual lower bound, since using `<` will stop if the search lands on the element itself.

While adding `#:rightmost?` works for now, I think I will replace this function with two funtions: `lower-bound` and an `upper-bound` .   I will, of course, update all uses of the function, but the main difference is that series will require strict ordering (i.e. using `<` instead of `<=` as the compare function).

You don't need to do anything for this PR, I am just letting you know that a backwards incompatible change will come to the data-frame package (you will need to use strict ordering for the series).

alex-hhh commented on private/test/df-test.rkt:
```diff
@@ -1037,7 +1085,7 @@
                       #:mode 'logarithmic
                       #:residual? #t))
-     (check-modified-residuals fit-log df "base2" "log")
+     (check-modified-residuals fit-log df "base2" "log"))))
```
Note that by moving the closing parens here, you just put the commented out code below out of the test suite.  While it does not changed the meaning of the program, it has the potential to introduce subtle problems....

ralsei replied:
This was my editor trying to be smarter than I was. My apologies.

alex-hhh commented on scribblings/data-frame.scrbl:
```diff
@@ -421,6 +437,16 @@ value, which is a list with all the values from the selected
+search is necessary.
+
+This does not check whether the series is actually sorted, rather whether it
+has been blessed with @racket[df-set-sorted!].
```
`df-set-sorted!` checks that the series is sorted before applying the flag, so the above sentence implies that `df-is-sorted?` might return `#t` for a series which is actually unsorted.

ralsei replied:
Ah, not the intent. I meant to say that a sorted series may not be blessed with `df-set-sorted!` -- I'll clarify the wording.

alex-hhh commented on scribblings/data-frame.scrbl:
```diff
@@ -421,6 +437,16 @@ value, which is a list with all the values from the selected
 }
 
+@defproc[(df-is-sorted? (df data-frame?) (series string?)) boolean?]{
+Determines if the given series @racket[series] in the data-frame @racket[df]
```
I think it would be clearer if the documentation would simply state:

> Return #t if the data-frame series is sorted (i.e. `df-set-sorted! has been called for this series).

alex-hhh commented on private/test/df-test.rkt:
```diff
@@ -544,21 +577,36 @@
+     (check equal? (df-index-of df "col7" 1) 0)
+     (check equal? (df-index-range df "col7" 1) (cons 0 1))
+     (check equal? (df-index-of df "col7" 2) 2)
+     (check equal? (df-index-range df "col7" 2) (cons 2 3))
```
Looking at this test, I think `df-index-range` might not be correctly implemented.  I would expect the range to be "open ended", i.e the end value is the first value which is not equal to the search term.  For the example above, I would expect that for the series [1 1 2 2], the equal-range for 2 is `(cons 2 4)` and for the value 1 to be `(cons 0 2)`

In particular, what should `df-index-range` return if the value is not found (i.e. you search for 1.9)?  Maybe add a test for this case as well?

Note that all the ranges (for example the `#:start` and `#:stop` ranges in `df-select`) in the data-frame package are open ended, and "end" points to "one after" the last value in the range.  This allows specifying an empty range at a certain position.

alex-hhh reviewed (changes requested):
Hi @ralsei, I had a first look at this PR, and while overall it looks good, there are some function name changes I would like you to make and also clarify the behavior of `df-index-range` in corner cases.

alex-hhh commented:
Another consideration you might want to give to `df-index-range` is whether it should return two values or a cons cell.  If you find that most of the uses of `df-index-range` are like so:

```racket
(match-define (cons start end) (df-index-range ...))
```

You might want to make `df-index-range` return two values instead, as this would be more efficient.

alex-hhh committed:
Replace bsearch with lower-bound and upper-bound functions

... since these have a more clear semantics than `bsearch` when multiple equal
values are present in the data.

alex-hhh commented:
Hi @ralsei , can you pull from the ah/lower-bound branch as this implements the `lower-bound` and `upper-bound` functions, plus an `equal-range` function which you can use in the `series-index-range` function?  These functions have better semantics than `bsearch`, but require the use of strict ordering for the sort function (e.g. use `<` instead of `<=`)

https://github.com/alex-hhh/data-frame/tree/ah/lower-bound

ralsei committed:
Merge remote-tracking branch 'upstream/ah/lower-bound' into issue-8

ralsei committed:
update usage of lower-bound and upper-bound

ralsei commented:
So the issue with the changes in the `ah/lower-bound` branch is that sorting with `<=` no longer produces expected behavior, _but_ using `check-valid-sort` with `<` and duplicate items in a series fails. This is why I added an `equal?` check -- but I don't know if this is what is wanted.

Still going to incorporate the requested changes.

ralsei committed:
series-shallow-copy -> copy-series

ralsei committed:
df-equal-range

ralsei committed:
series-equal-range, not in series

ralsei commented:
I think I have resolved all requested changes. Feel free to tell me if it doesn't.

alex-hhh reviewed (approved):
The changes look good, so I will merge this now.
